fdbrpc/HTTP.actor.cpp

```diff
@@ -525,4 +525,135 @@ ACTOR Future<Reference<HTTP::Response>> doRequest(Reference<IConnection> conn,
 	}
 }
 
+ACTOR Future<Void> sendProxyConnectRequest(Reference<IConnection> conn,
+                                           std::string remoteHost,
+                                           std::string remoteService) {
+	state Headers headers;
+	headers["Host"] = remoteHost + ":" + remoteService;
+	headers["Accept"] = "application/xml";
+	headers["Proxy-Connection"] = "Keep-Alive";
+	state int requestTimeout = 60;
+	state int maxTries = FLOW_KNOBS->RESTCLIENT_CONNECT_TRIES;
+	state int thisTry = 1;
+	state double nextRetryDelay = 2.0;
+	state Reference<IRateControl> sendReceiveRate = makeReference<Unlimited>();
+	state int64_t bytes_sent = 0;
+
+	loop {
+		state Optional<Error> err;
+		state Reference<Response> r;
+
+		try {
+			Reference<Response> _r = wait(timeoutError(doRequest(conn,
+			                                                     "CONNECT",
+			                                                     remoteHost + ":" + remoteService,
+			                                                     headers,
+			                                                     nullptr,
+			                                                     0,
+			                                                     sendReceiveRate,
+			                                                     &bytes_sent,
+			                                                     sendReceiveRate),
+			                                           requestTimeout));
+			r = _r;
+		} catch (Error& e) {
+			if (e.code() == error_code_actor_cancelled)
+				throw;
+			err = e;
+		}
+
+		// If err is not present then r is valid.
+		// If r->code is in successCodes then record the successful request and return r.
+		if (!err.present() && r->code == 200) {
+			return Void();
+		}
+
+		// All errors in err are potentially retryable as well as certain HTTP response codes...
+		bool retryable = err.present() || r->code == 500 || r->code == 502 || r->code == 503 || r->code == 429;
+
+		// But only if our previous attempt was not the last allowable try.
+		retryable = retryable && (thisTry < maxTries);
+
+		TraceEvent event(SevWarn, retryable ? "ProxyConnectCommandFailedRetryable" : "ProxyConnectCommandFailed");
+
+		// Attach err to trace event if present, otherwise extract some stuff from the response
+		if (err.present()) {
+			event.errorUnsuppressed(err.get());
+		}
+		event.suppressFor(60);
+		if (!err.present()) {
+			event.detail("ResponseCode", r->code);
+		}
+
+		event.detail("ThisTry", thisTry);
+
+		// If r is not valid or not code 429 then increment the try count.  429's will not count against the attempt
```

Comment:
What is code 429?

Reply:
429 is Too Many Requests. I basically copied everything from https://github.com/apple/foundationdb/blob/a8dd6b65ec03df7ba0dba76eaf5964766142500e/fdbclient/S3BlobStore.actor.cpp#L963